File: app/src/main/java/com/example/studybuddy/mlc/TVMBridge.java
package com.example.studybuddy.mlc;

import android.content.Context;
import android.util.Log;

/**
 * Bridge class for working with MLC-LLM using JNI
 */
public class TVMBridge {
    private static final String TAG = "TVMBridge";
    private static boolean librariesLoaded = false;
    
    static {
        try {
            // Load the required libraries in the correct order
            System.loadLibrary("c++_shared");
            System.loadLibrary("tvm_runtime");
            System.loadLibrary("tvm");
            System.loadLibrary("mlc_llm");
            System.loadLibrary("mlc_llm_module");
            
            // Load our JNI bridge library that implements the native methods
            System.loadLibrary("tvm_bridge");
            
            librariesLoaded = true;
            Log.i(TAG, "TVM libraries loaded successfully");
        } catch (UnsatisfiedLinkError e) {
            Log.e(TAG, "Failed to load TVM libraries: " + e.getMessage());
        }
    }
    
    // Check if libraries were loaded successfully
    public static boolean areLibrariesLoaded() {
        return librariesLoaded;
    }
    
    // Initialize the MLC-LLM runtime
    public static native boolean initRuntime(String modelPath);
    
    // Generate text response for a prompt
    public static native String generateText(String prompt, int maxTokens);
    
    // Set various generation parameters
    public static native boolean setTemperature(float temperature);
    public static native boolean setTopP(float topP);
    public static native boolean setRepetitionPenalty(float penalty);
    
    // Clean up resources
    public static native void destroyRuntime();
    
    /**
     * Interface for streaming text generation callbacks
     */
    public interface StreamingCallback {
        /**
         * Called when a new token is generated
         * @param token The token text
         * @param isFinished True if generation is complete, false if more tokens are coming
         */
        void onToken(String token, boolean isFinished);
    }
    
    /**
     * Start streaming text generation
     * @param prompt The input prompt
     * @param maxTokens Maximum number of tokens to generate
     * @param callback Callback to receive generated tokens
     * @return True if generation started successfully
     */
    public static native boolean startStreamingGeneration(String prompt, int maxTokens, StreamingCallback callback);
    
    /**
     * Stop any ongoing streaming generation
     */
    public static native void stopStreamingGeneration();
} 

File: app/src/main/java/com/example/studybuddy/ml/TVMBridge.kt
package com.example.studybuddy.ml

import android.util.Log
import java.io.File

/**
 * Bridge to the native TVM/MLC runtime.
 */
class TVMBridge {
    companion object {
        private const val TAG = "TVMBridge"
        
        init {
            try {
                System.loadLibrary("c++_shared")
                System.loadLibrary("tvm_runtime")
                System.loadLibrary("tvm")
                System.loadLibrary("mlc_llm")
                System.loadLibrary("mlc_llm_module")
                System.loadLibrary("tvm_bridge")
                Log.i(TAG, "TVM libraries loaded successfully")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load native libraries: ${e.message}")
                throw RuntimeException("Failed to load required native libraries: ${e.message}")
            }
        }
    }
    
    /**
     * Initialize the TVM engine with the given model path
     */
    fun initializeEngine(modelPath: String): Boolean {
        Log.i(TAG, "Initializing MLC-LLM runtime with model at: $modelPath")
        
        // List the contents of the model directory for debugging
        val modelDir = File(modelPath)
        if (modelDir.exists()) {
            Log.d(TAG, "Directory contents of $modelPath:")
            modelDir.listFiles()?.forEach { file ->
                Log.d(TAG, "  ${file.name}")
            }
            
            // Count parameter files
            val paramFiles = modelDir.listFiles { file -> file.name.startsWith("params_shard_") }
            Log.d(TAG, "Found ${paramFiles?.size ?: 0} parameter files")
        } else {
            val errorMsg = "Model directory does not exist: $modelPath"
            Log.e(TAG, errorMsg)
            return false
        }
        
        // Call the C++ implementation to initialize
        return try {
            val result = initializeTVMRuntime(modelPath)
            if (!result) {
                Log.e(TAG, "Native TVM runtime initialization failed")
                return false
            }
            Log.i(TAG, "Native TVM runtime initialized successfully")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Exception initializing native TVM runtime: ${e.message}")
            false
        }
    }
    
    /**
     * Generate a response for a prompt using non-streaming mode
     */
    fun chat(prompt: String): String {
        Log.d(TAG, "Generating response for prompt: $prompt")
        val response = generateResponse(prompt)
        if (response.startsWith("ERROR:") || response.startsWith("Error:")) {
            Log.e(TAG, "Error in text generation: $response")
            throw RuntimeException(response)
        }
        return response
    }
    
    /**
     * Stream tokens from the model for the given prompt
     */
    fun streamChat(prompt: String, callback: (String) -> Unit) {
        Log.d(TAG, "Streaming response for prompt: $prompt")
        streamResponse(prompt) { token ->
            if (token.startsWith("ERROR:") || token.startsWith("Error:")) {
                Log.e(TAG, "Error in streaming: $token")
                throw RuntimeException(token)
            }
            callback(token)
        }
    }
    
    /**
     * Set temperature for text generation
     */
    fun setTemperature(temperature: Float) {
        Log.i(TAG, "Temperature set to: $temperature")
        setGenerationTemperature(temperature)
    }
    
    /**
     * Set top_p for text generation
     */
    fun setTopP(topP: Float) {
        Log.i(TAG, "Top-p set to: $topP")
        setGenerationTopP(topP)
    }
    
    /**
     * Reset the chat session
     */
    fun resetChat() {
        Log.d(TAG, "Resetting chat session")
        resetChatSession()
    }
    
    // Native method declarations
    private external fun initializeTVMRuntime(modelPath: String): Boolean
    private external fun generateResponse(prompt: String): String
    private external fun streamResponse(prompt: String, callback: (String) -> Unit)
    private external fun setGenerationTemperature(temperature: Float): Boolean
    private external fun setGenerationTopP(topP: Float): Boolean
    private external fun resetChatSession(): Boolean
} 